### ISSUE TITLE
Preview/Thumbnail shown on Map

### ISSUE THREAD

**Issue by m-mohr** [stac-utils/stac-layer#9]
I think the desired behavior that only overviews are shown on the map, but not previews (from links) or thumbnails (from assets). Right now it shows me one of them on the map...

**Comment by m-mohr** [comment]
Yeah, there's an issue in the code:
https://github.com/stac-utils/stac-layer/blob/main/src/index.js#L542

It seems right now that stay-layer looks for an asset with key thumbnail, but it should look for an asset with role overview.

**Comment by m-mohr** [comment]
PR: #10

**Comment by DanielJDufour** [comment]
That's my mistake.  I forgot to add support for the `displayOverview` option.  I see PR #10 and will look to merge that work in soon.  However, there's one point of confusion.  I was under the impression we did want to give people the option on whether or not they wanted to display preview/thumbnail assets through the `displayPreview` option and stac-browser would pass in `displayPreview: false` when calling `stacLayer`.

Would it be perhaps okay to keep `displayPreview` as an option but set the default to false while of course also adding support for `displayOverview`?

Additionally, do you happen to have any good test cases where the STAC item includes an overview asset?

**Comment by m-mohr** [comment]
Yes, your alternative is also fine or even better. My PR was just the easiest solution to quickly make it work. Feel free to update the PR (or close and create a new one) to support the other approach.

No, unfortunately not out of my head. I thought Planet would have it, but they use visual for what should be overview, I think.

**Comment by DanielJDufour** [comment]
You read my mind!  I was just working on this earlier today :-) I think the next step will be to create some sample test cases.  Once I'm done with those, I'll share those with you and see if that covers the remaining issues.  I'll then merge your PR into a branch and build off of that to code up a solution.

**Comment by DanielJDufour** [comment]
Hey, @m-mohr .  I created this PR here to track updates to displayOverview functionality
https://github.com/stac-utils/stac-layer/pull/13

I added a test case which you can check here: https://github.com/stac-utils/stac-layer/blob/8c0328bc6fd29d6a63d2d4b050edf822d8462225/src/test/stac-item/cogs/cog-no-cors-overview.html#L35

I created a sample overview PNG file of the Canopy Base Height COG for California using https://observablehq.com/@danieljdufour/convert-geotiff-to-png-using-georaster.

It's small so I checked it in here: https://github.com/stac-utils/stac-layer/blob/8c0328bc6fd29d6a63d2d4b050edf822d8462225/src/test/stac-item/cogs/California-Vegetation-CanopyBaseHeight-2016-Summer-00010m.png

Let me know if you think added a PNG Overview asset is a good test case.

Thank you!

**Comment by DanielJDufour** [comment]
okay, cool.  I'll add another test case for the cog.

**Comment by DanielJDufour** [comment]
I created an external overview file using [gdaladdo](https://gdal.org/programs/gdaladdo.html) and it appears that the overview file doesn't include projection info like CRS/SRS.  (I think because it assumes that the original file is available).  This complicates things, but I think there are workarounds.

There's several different cases the code needs to gracefully handle.  I would like your help determining which cases we are required to support.

## 1) Overview is a JPG/PNG file and is accessible.
- Possible Solution: Simply use the STAC Item's BBox and Leaflet's ImageLayer

## 2) Overview is a COG with geospatial info (including projection)
- Possible Solution: use GeoRaster/TiTiler to try to display it

## 3) Both Original GeoTIFF and Overview are Accessible
- Good news is that GeoRaster automatically looks for a file with the same name + ".ovr".  It doesn't really need an overview file, but will use it when it makes sense (zoomed out).
- Bad news is that GeoRaster currently doesn't support a way for a user to pass in an overview url
- Possible Solution: Change GeoRaster to allow people to specify an overview url (but this will require some additional coding)

## 4) Original GeoTIFF is inaccessible (sometimes CORS or it's too large) and Overview is Accessible but without geospatial information
- Not sure on solution.  But could use georaster or maybe geotiff.js to convert the COG to a JPG/PNG and then use Leaflet's ImageLayer.  (Example using georaster: https://observablehq.com/@danieljdufour/convert-geotiff-to-png-using-georaster).  One problem will be that georaster throws an error if no projection info is present.  I think I could hack together a solution using geotiff.js to read the values from a low resolution internal image and then calling parseGeoRaster with an in-memory object like [here](https://github.com/geotiff/georaster#load-from-simple-object-on-front-end).  The code won't look nice, but it **should** work.

As always, open to your thoughts and advice.

**Comment by DanielJDufour** [comment]
@m-mohr , great comments.  Thank you.

Would "Displayable Assets" only include those with RGB/RGBA bands or a color palette?  Is this synonymous with "visual"?  Would it also include single-band GeoTIFFs that can be displayed on a dynamically calculated linear color scale (like https://twitter.com/DanielJDufour/status/1431300791319638022)? I'm assuming no, but want to make sure I understand.

Another question I have is, could you expand on your preference to display the overview before the others?  Is it because it would put less strain on the browser's CPU / memory?  Not disagreeing with the idea, just want to better understand how we sort priorities (visualization vs. performance vs. accuracy, etc.).

Regarding displayOrder, it's certainly desirable to give users that level of customization.  But I need some time to think how to implement it.  The current code makes a bunch of assumptions about display order and it might be tricky to find a way to model them concisely in an array.  For example, the code currently looks to see if the RGB bands are in separate GeoTIFFs (like Landsat 8 on AWS) if there are multiple assets and no asset with the "visual" key.  I'm not sure how to represent this as an array option.  Maybe as the string "separated" like `["overview", "separated", "visual", "data"] `?  Or we could do something more complicated like
```js
[{ role: "overview" }, [{ common_name: "red" }, { common_name: "green" }, { common_name: "blue" }], { name: /visual/i }, { name: /data/i }]
```

**Comment by m-mohr** [comment]
> Would "Displayable Assets" only include those with RGB/RGBA bands or a color palette?

For all formats except GeoTiff: yes
For GeoTiff: No, I'd expect that overview and visual only have RGBA (or a single/grayscale band), but as it may also display the normal data assets, it may also be something else.

> Is this synonymous with "visual"?
overview and visual are very similar usually, but otherwise no (see above).

> Would it also include single-band GeoTIFFs

Yes. I mean overall in STAC we just gives recommendations what people should provide for these roles, but in the end they could provide anything (even a 2000 band image as a preview if they so desire, but we focus on common use cases here and don't need to support every exotic use case out there).

> Another question I have is, could you expand on your preference to display the overview before the others? Is it because it would put less strain on the browser's CPU / memory?

Yes and it's basically the order that we had in mind when writing STAC. Usually the overview (and the non-standardized[!] visual assets) should optimized vor visualization purposes (e.g. in the browser) and as such I'd expect better performance overall. Right now my browser is already very slow if I open the Planet COGS with 256 resolution. I'm not sure whether it's due to WebGL, but the COGs render much faster on my machine with OpenLayers compared to georaster/leaflet. And if I make them equally fast (by tuning the resolution option), I get very pixelated images.

> Regarding displayOrder, it's certainly desirable to give users that level of customization. But I need some time to think how to implement it.

We can start without it and leave this for a later version for sure, but it would be nice to get the overviews in. :-)

>  For example, the code currently looks to see if the RGB bands are in separate GeoTIFFs (like Landsat 8 on AWS) if there are multiple assets and no asset with the "visual" key. 

I'm not sure this is the way to go anyway. The roles are really independant of the bands. The code should look at the eo:bands/raster:bands to see how many and which bands are available. If you don't get enough details about it from the metadata and the image is not RGB(A), I'd simply show a single band as otherwise we are really just guessing.

> Maybe as the string "separated" like `["overview", "separated", "visual", "data"] `?

That doesn't really work as it's not part of the STAC spec and at least STAC Browser is not configured for specific COG "types" by default, so I separated would make an assumption that in most cases is not true.

> Or we could do something more complicated 

Let's don't make it too complicated, overall this was just meant to be a "bugfix" for an oversight and not something completely new (although I appreciate your thoughts and ideas). :-)

**Comment by DanielJDufour** [comment]
Sounds good.  I appreciate the thoughtful feedback.  I'll get the overview support in and remove the unreliable assumptions about separated RGB bands.

There's a few factors why the COGs render more quickly for you using OL vs. GeoRasterLayer.  The most important factor is probably that OL is using WebGL whereas GeoRasterLayer is using the CPU.  But I'd be curious to test the speed difference once [geowarp](https://github.com/danieljdufour/geowarp) is added as a rendering option to GeoRasterLayer.
